lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -396,94 +410,173 @@ abstract protected function hydrateAllData();
      * @psalm-return array{
+     *                   ids: array<array-key, array<array-key, mixed>>,
      *                   data: array<array-key, array>,
      *                   newObjects?: array<array-key, array{
      *                       class: mixed,
-     *                       args?: array
+     *                       args: array
      *                   }>,
      *                   scalars?: array
      *               }
      */
     protected function gatherRowData(array $data, array &$id, array &$nonemptyComponents)
     {
-        $rowData = ['data' => []];
+        $rowData = [
+            'ids'  => $this->gatherRowIDs($data, $id, $nonemptyComponents),
+            'data' => [],
+        ];
 
-        foreach ($data as $key => $value) {
-            $cacheKeyInfo = $this->hydrateColumnInfo($key);
-            if ($cacheKeyInfo === null) {
-                continue;
+        foreach (array_keys($this->_rowMappingInfo['dql']) as $alias) {
+            $rowData['data'][$alias] = $this->gatherRowDataFromDQLMapping($data, $alias);
+        }
+
+        if (isset($this->_rowMappingInfo['scalars'])) {
+            $rowData['scalars'] = $this->gatherRowDataFromScalarsMapping($data);
+        }
+
+        if (isset($this->_rowMappingInfo['newObjects'])) {
+            $rowData['newObjects'] = $this->gatherRowDataFromNewObjectsMapping($data);
+        }
+
+        return $rowData;
+    }
+
+    /**
+     * @param mixed[] $data SQL Result Row.
+     * @psalm-param array<array-key, string> $id                 Dql-Alias => ID-Hash.
+     * @psalm-param array<array-key, bool>   $nonemptyComponents Does this DQL-Alias has at least one non NULL value?
+     *
+     * @psalm-return array<array-key, array<array-key, mixed>> An array with all the ids (name => value) of the row.
+     */
+    protected function gatherRowIDs(array $data, array &$id, array &$nonemptyComponents)
+    {
+        $ids = [];
+
+        foreach ($this->_rowMappingInfo['ids'] as $column => $cacheKeyInfo) {
+            $dqlAlias  = $cacheKeyInfo['dqlAlias'];
+            $fieldName = $cacheKeyInfo['fieldName'];
+
+            $ids[$dqlAlias][$fieldName] = $data[$column];
+
+            if ($data[$column] !== null) {
+                $id[$dqlAlias]                .= '|' . $data[$column];
+                $nonemptyComponents[$dqlAlias] = true;
             }
+        }
+
+        return $ids;
+    }
+
+    /**
+     * @param mixed[] $data SQL Result Row.
+     * @psalm-param string $dqlAlias                 Dql-Alias
+     *
+     * @psalm-return array<array-key, array> An array with all the fields
+     *                                             (name => value) of the data
+     *                                             row, grouped by their
+     *                                             component alias.
+     */
+    protected function gatherRowDataFromDQLMapping(array $data, string $dqlAlias)
```

Comment:
Does this method have to be protected? And please use native return types on new methods if you can.

```suggestion
    private function gatherRowDataFromDQLMapping(array $data, string $dqlAlias): array
```

Reply:
this method (and the other ones introduced here) are made protected to be callable from the Object Hydrator

Reply:
(return types added)